File: app/services/tenant_provisioning.py
"""Aprovisionamiento de un tenant nuevo.

Al crear un tenant hay que sembrar su configuracion minima para operar:
estados de pedido, un rol admin con todos los permisos, y un impuesto por
defecto. El flujo de estados varia segun `business_type`; el rol y el
impuesto no. Si un modelo de negocio nuevo exige tocar codigo aqui en vez
de agregar un preset, el diseño esta fallando (ver docs/multi-tenancy.md).
"""

from decimal import Decimal

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.domain.tenant_settings import defaults_for, validate
from app.models.order_status import OrderStatus, OrderStatusTransition
from app.models.permission import Permission
from app.models.role import Role
from app.models.tax_rate import TaxRate
from app.models.tenant import Tenant
from app.repositories import tenant_repository

# code, name, category, sort_order, is_initial, is_final
_StatusDef = tuple[str, str, str, int, bool, bool]
# from_code, to_code, required_permission
_TransitionDef = tuple[str, str, str | None]

STATUS_PRESETS: dict[str, dict[str, list]] = {
    "fast_food": {
        "statuses": [
            ("pending", "Pendiente", "new", 1, True, False),
            ("paid", "Pagado", "new", 2, False, False),
            ("preparing", "Preparacion", "kitchen", 3, False, False),
            ("ready", "Listo", "ready", 4, False, False),
            ("delivered", "Entregado", "completed", 5, False, True),
            ("cancelled", "Cancelado", "cancelled", 9, False, True),
        ],
        "transitions": [
            ("pending", "paid", "payments.register"),
            ("paid", "preparing", "orders.advance_kitchen"),
            ("preparing", "ready", "orders.advance_kitchen"),
            ("ready", "delivered", "orders.advance_kitchen"),
            ("pending", "cancelled", "orders.cancel"),
            ("paid", "cancelled", "orders.cancel"),
        ],
    },
    "table_service": {
        "statuses": [
            ("open", "Abierto", "new", 1, True, False),
            ("preparing", "Preparacion", "kitchen", 2, False, False),
            ("ready", "Listo", "ready", 3, False, False),
            ("served", "Servido", "completed", 4, False, True),
            ("cancelled", "Cancelado", "cancelled", 9, False, True),
        ],
        "transitions": [
            ("open", "preparing", "orders.advance_kitchen"),
            ("preparing", "ready", "orders.advance_kitchen"),
            ("ready", "served", "orders.advance_kitchen"),
            ("open", "cancelled", "orders.cancel"),
            ("preparing", "cancelled", "orders.cancel"),
        ],
    },
    "delivery": {
        "statuses": [
            ("pending", "Pendiente", "new", 1, True, False),
            ("paid", "Pagado", "new", 2, False, False),
            ("preparing", "Preparacion", "kitchen", 3, False, False),
            ("ready", "Listo", "ready", 4, False, False),
            ("in_transit", "En camino", "in_transit", 5, False, False),
            ("delivered", "Entregado", "completed", 6, False, True),
            ("cancelled", "Cancelado", "cancelled", 9, False, True),
        ],
        "transitions": [
            ("pending", "paid", "payments.register"),
            ("paid", "preparing", "orders.advance_kitchen"),
            ("preparing", "ready", "orders.advance_kitchen"),
            ("ready", "in_transit", "delivery.assign"),
            ("in_transit", "delivered", "delivery.complete"),
            ("pending", "cancelled", "orders.cancel"),
            ("paid", "cancelled", "orders.cancel"),
        ],
    },
}


def provision_tenant(db: Session, tenant: Tenant) -> None:
    """Siembra estados, rol admin e impuesto por defecto para un tenant recien creado."""
    preset = STATUS_PRESETS.get(tenant.business_type, STATUS_PRESETS["fast_food"])

    status_by_code: dict[str, OrderStatus] = {}
    for code, name, category, sort_order, is_initial, is_final in preset["statuses"]:
        status = OrderStatus(
            tenant_id=tenant.id,
            code=code,
            name=name,
            category=category,
            sort_order=sort_order,
            is_initial=is_initial,
            is_final=is_final,
        )
        db.add(status)
        status_by_code[code] = status
    db.flush()  # asigna ids antes de crear las transiciones

    for from_code, to_code, permission in preset["transitions"]:
        db.add(
            OrderStatusTransition(
                from_status_id=status_by_code[from_code].id,
                to_status_id=status_by_code[to_code].id,
                required_permission=permission,
            )
        )

    db.add(
        TaxRate(
            tenant_id=tenant.id,
            name="Impuesto general",
            rate=Decimal("0"),
            included_in_price=True,
            is_default=True,
        )
    )

    admin_role = Role(tenant_id=tenant.id, code="admin", name="Administrador", is_system=True)
    admin_role.permissions = list(db.scalars(select(Permission)).all())
    db.add(admin_role)

    db.flush()


def create_tenant(
    db: Session, *, name: str, business_type: str = "fast_food", currency: str = "COP", settings: dict | None = None
) -> Tenant:
    """Crea un tenant y lo deja listo para operar: aplica alta + aprovisionamiento en una sola transaccion."""
    if settings is None:
        settings = defaults_for(business_type)
    else:
        validate(settings)

    tenant = tenant_repository.create(db, name=name, business_type=business_type, currency=currency, settings=settings)
    provision_tenant(db, tenant)
    db.commit()
    db.refresh(tenant)
    return tenant

File: php/src/Services/TenantProvisioning.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Core\Database;
use App\Core\Permissions;
use App\Domain\SettingsError;
use App\Domain\Slug;
use App\Domain\TenantSettings;
use App\Models\Tenant;
use App\Repositories\DiscountReasonRepository;
use App\Repositories\OrderStatusRepository;
use App\Repositories\RoleRepository;
use App\Repositories\TaxRateRepository;
use App\Repositories\TenantRepository;
use PDO;

final class TenantProvisioning
{
    /**
     * [code, name, category, sort_order, is_initial, is_final][]
     * [from_code, to_code, required_permission|null][]
     *
     * @var array<string, array{statuses: array<int, array{0:string,1:string,2:string,3:int,4:bool,5:bool}>, transitions: array<int, array{0:string,1:string,2:?string}>}>
     */
    private const STATUS_PRESETS = [
        'fast_food' => [
            'statuses' => [
                ['pending', 'Pendiente', 'new', 1, true, false],
                ['paid', 'Pagado', 'new', 2, false, false],
                ['preparing', 'Preparacion', 'kitchen', 3, false, false],
                ['ready', 'Listo', 'ready', 4, false, false],
                ['delivered', 'Entregado', 'completed', 5, false, true],
                ['cancelled', 'Cancelado', 'cancelled', 9, false, true],
            ],
            'transitions' => [
                ['pending', 'paid', 'payments.register'],
                ['paid', 'preparing', 'orders.advance_kitchen'],
                ['preparing', 'ready', 'orders.advance_kitchen'],
                ['ready', 'delivered', 'orders.advance_kitchen'],
                ['pending', 'cancelled', 'orders.cancel'],
                ['paid', 'cancelled', 'orders.cancel'],
            ],
        ],
        'table_service' => [
            'statuses' => [
                ['open', 'Abierto', 'new', 1, true, false],
                ['preparing', 'Preparacion', 'kitchen', 2, false, false],
                ['ready', 'Listo', 'ready', 3, false, false],
                ['served', 'Servido', 'completed', 4, false, true],
                ['cancelled', 'Cancelado', 'cancelled', 9, false, true],
            ],
            'transitions' => [
                ['open', 'preparing', 'orders.advance_kitchen'],
                ['preparing', 'ready', 'orders.advance_kitchen'],
                ['ready', 'served', 'orders.advance_kitchen'],
                ['open', 'cancelled', 'orders.cancel'],
                ['preparing', 'cancelled', 'orders.cancel'],
            ],
        ],
        'delivery' => [
            'statuses' => [
                ['pending', 'Pendiente', 'new', 1, true, false],
                ['paid', 'Pagado', 'new', 2, false, false],
                ['preparing', 'Preparacion', 'kitchen', 3, false, false],
                ['ready', 'Listo', 'ready', 4, false, false],
                ['in_transit', 'En camino', 'in_transit', 5, false, false],
                ['delivered', 'Entregado', 'completed', 6, false, true],
                ['cancelled', 'Cancelado', 'cancelled', 9, false, true],
            ],
            'transitions' => [
                ['pending', 'paid', 'payments.register'],
                ['paid', 'preparing', 'orders.advance_kitchen'],
                ['preparing', 'ready', 'orders.advance_kitchen'],
                ['ready', 'in_transit', 'delivery.assign'],
                ['in_transit', 'delivered', 'delivery.complete'],
                ['pending', 'cancelled', 'orders.cancel'],
                ['paid', 'cancelled', 'orders.cancel'],
            ],
        ],
    ];

    /** Siembra estados, rol admin e impuesto por defecto para un tenant recien creado. */
    /** Los motivos con los que arranca cualquier restaurante; despues los edita. */
    private const MOTIVOS_DESCUENTO = ['Cortesía', 'Reclamo del cliente', 'Empleado', 'Convenio'];

    public static function provisionTenant(PDO $pdo, string $tenantId, string $businessType): void
    {
        $preset = self::STATUS_PRESETS[$businessType] ?? self::STATUS_PRESETS['fast_food'];
        $statusRepo = new OrderStatusRepository($pdo);

        $idByCode = [];
        foreach ($preset['statuses'] as [$code, $name, $category, $sortOrder, $isInitial, $isFinal]) {
            $idByCode[$code] = $statusRepo->insertStatus($tenantId, $code, $name, $category, $sortOrder, $isInitial, $isFinal);
        }

        foreach ($preset['transitions'] as [$fromCode, $toCode, $permission]) {
            $statusRepo->insertTransition($idByCode[$fromCode], $idByCode[$toCode], $permission);
        }

        (new TaxRateRepository($pdo))->create($tenantId, 'Impuesto general', '0', true, true);

        // Los motivos de descuento que aparecen en cualquier restaurante. Sin
        // ninguno, el primer descuento seria imposible de aplicar: el dominio
        // exige motivo y la lista estaria vacia.
        $reasons = new DiscountReasonRepository($pdo);
        foreach (self::MOTIVOS_DESCUENTO as $orden => $motivo) {
            $reasons->create($tenantId, $motivo, $orden);
        }

        $roles = new RoleRepository($pdo);
        // is_system=true: es la salida de emergencia del tenant (ver
        // AdminService::setRolePermissions), asi que no puede quedar como un
        // rol mas que alguien edite hasta dejarlo sin permisos.
        $adminRole = $roles->create($tenantId, 'admin', 'Administrador', isSystem: true);
        // Todos los del catalogo de la plataforma, que desde F5.5 es la
        // fuente: antes se leian de la tabla `permissions`, que es el destino
        // del join y no el catalogo.
        $roles->setPermissions($adminRole->id, Permissions::codes());
    }

    /**
     * Crea un tenant y lo deja listo para operar: alta + aprovisionamiento
     * en una sola transaccion.
     *
     * @param array<mixed>|null $settings
     */
    public static function createTenant(
        string $name,
        string $businessType = 'fast_food',
        string $currency = 'COP',
        ?array $settings = null,
        ?string $slug = null,
    ): Tenant {
        if ($settings === null) {
            $settings = TenantSettings::defaultsFor($businessType);
        } else {
            try {
                TenantSettings::validate($settings);
            } catch (SettingsError $e) {
                throw new TenantProvisioningError($e->getMessage(), previous: $e);
            }
        }

        // Si quien llama ya abrio una transaccion (el script de alta, que
        // ademas crea la primera sucursal y el primer usuario), se respeta:
        // el alta completa tiene que ser todo o nada, no dejar a medias un
        // restaurante sin usuario con el cual entrar.
        $pdo = Database::admin();
        $ownsTransaction = !$pdo->inTransaction();
        if ($ownsTransaction) {
            $pdo->beginTransaction();
        }

        try {
            // El slug se genera del nombre si no lo dan: es lo que alguien
            // escribe al entrar cuando su correo esta en mas de un
            // restaurante, y una cosa menos que inventar al dar de alta.
            $tenants = new TenantRepository($pdo);
            $elegido = Slug::unique($slug ?? $name, static fn (string $s) => $tenants->slugExists($s));
            $tenant = $tenants->create($name, $elegido, $businessType, $currency, $settings);
            self::provisionTenant($pdo, $tenant->id, $businessType);
            if ($ownsTransaction) {
                $pdo->commit();
            }
            return $tenant;
        } catch (\Throwable $e) {
            if ($ownsTransaction) {
                $pdo->rollBack();
            }
            throw $e;
        }
    }
}
